ui/src/routes/tfe/$.tsx

```diff
@@ -52,60 +55,76 @@ async function handler({ request }) {
-    
-    // Get email from statesman (token service doesn't store email)
+  
+  // Check cache first
+  const cached = tokenCache.get(token);
+  if (cached) {
+    userId = cached.userId;
+    userEmail = cached.userEmail;
+    orgId = cached.orgId;
+    console.log(`✅ Token cache hit for ${userId}`);
+  } else {
+    // Cache miss - verify token
     try {
-      const userResponse = await fetch(`${process.env.STATESMAN_BACKEND_URL}/internal/api/users/${userId}`, {
-        headers: {
-          'Authorization': `Bearer ${process.env.STATESMAN_BACKEND_WEBHOOK_SECRET}`,
-          'X-Org-ID': orgId,
-          'X-User-ID': userId,
-          'X-Email': '',
-        },
+      const startVerify = Date.now();
+      
+      const verifyResponse = await fetch(`${process.env.TOKENS_SERVICE_BACKEND_URL}/api/v1/tokens/verify`, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ token }),
       });
       
-      if (userResponse.ok) {
-        const userData = await userResponse.json();
-        userEmail = userData.email || '';
-        console.log('Got user email from statesman:', userEmail);
-      } else {
-        console.warn('Could not fetch user from statesman:', userResponse.status);
-        userEmail = '';
+      if (!verifyResponse.ok) {
+        console.error('Token verification failed:', verifyResponse.status);
+        return new Response('Unauthorized: Invalid token', { status: 401 })
       }
+      
+      const tokenInfo = await verifyResponse.json();
+      const verifyTime = Date.now() - startVerify;
+      
+      if (!tokenInfo.valid) {
+        return new Response('Unauthorized: Invalid token', { status: 401 })
+      }
+      
+      // Extract user info from token service response
+      const tokenData = tokenInfo.token || {};
+      userId = tokenData.user_id || tokenInfo.user_id || 'anonymous';
+      orgId = tokenData.org_id || tokenInfo.org_id || 'default';
+      userEmail = tokenData.email || tokenInfo.email || '';
+      
+      // Only fetch email if not in token AND if we need it
+      if (!userEmail) {
+        try {
+          const startEmail = Date.now();
+          const emailResponse = await fetch(`${process.env.STATESMAN_BACKEND_URL}/internal/api/users/${userId}`, {
+            headers: {
+              'Authorization': `Bearer ${process.env.STATESMAN_BACKEND_WEBHOOK_SECRET}`,
+              'X-Org-ID': orgId,
+              'X-User-ID': userId,
+              'X-Email': '',
+            },
+          });
+          
+          if (emailResponse.ok) {
+            const userData = await emailResponse.json();
+            userEmail = userData.email || '';
+            console.log(`Email lookup: ${Date.now() - startEmail}ms`);
+          } else {
+            userEmail = '';
+          }
+        } catch (error) {
+          console.error('Error fetching user email:', error);
+          userEmail = '';
+        }
+      }
```

Comment:
nit: this should go as a function in api/statesman_users.ts

Reply:
definitely makes it cleaner